shared/js/l10n.js

```diff
@@ -1172,10 +1175,22 @@
 
       switch (type) {
         case 'json':
-          io.loadJSON(path, onJSONLoaded, sync);
+          if (source === 'extra') {
+            navigator.mozApps.getLocalizationResource(
+              this.id, gaiaVersion, path, 'json')
+                .then(onJSONLoaded.bind(this, null), onL10nLoaded);
+          } else {
+            io.loadJSON(path, onJSONLoaded, sync);
+          }
           break;
         case 'properties':
-          io.load(path, onPropLoaded, sync);
+          if (source === 'extra') {
```

Comment:
There's a little bit of code duplication between `case 'json'` and `case 'properties'`, with the `source` check and calling two different IO methods.  My suggestion would be to create an object like so:

```
var bindingsIO = {
  extra: function(id, ver, path, type, callback, errback) {
    navigator.mozApps.getLocalizationResource(id, ver, type).then (callback, errback));
  },
  app: function(id, ver, path, type, callback, errback, sync) {
    // check type here
    io.loadJSON(path, callback, sync);
  },
};
```
…and then use it like so:

```
case 'json':
  bindingsIO[source](this.id, gaiaVersion, path, type, onJSONLoaded.bind(this, null), onL10nLoaded);
  break;
```